File: src/main/java/edu/princeton/cs/algs4/leetcode/exercise/SimplifyPath.java
package edu.princeton.cs.algs4.leetcode.exercise;

import edu.princeton.cs.algs4.leetcode.Stack;

import static edu.princeton.cs.algs4.StdOut.println;

/**
 * @author TuFei
 * @date 18-11-23.
 */
class SimplifyPath {
    public String simplifyPath(String path) {
        Stack stack = new Stack();
        StringBuilder temp = new StringBuilder();
        for (int i = 0; i < path.length(); i++) {
            char c = path.charAt(i);
            if (temp.length() == 0) {
                temp.append(c);
                continue;
            }
            if (c != '/') {
                temp.append(c);
            } else {
                switch (temp.toString()) {
                    case "/.":
                    case "/":
                        break;
                    case "/..":
                        if (!stack.isEmpty()) {
                            stack.pop();
                        }
                        break;
                    default:
                        stack.push(temp.toString());
                        break;
                }
                temp = new StringBuilder("/");
            }
        }

        String s = temp.toString();
        if ("/..".equals(s) && !stack.isEmpty()) {
            stack.pop();
        }
        if (s.length() != 0 && !"/".equals(s) && !"/.".equals(s) && !"/..".equals(s)) {
            stack.push(s);
        }
        if (stack.isEmpty()) {
            return "/";
        } else {
            StringBuilder result = new StringBuilder();
            while (!stack.isEmpty()) {
                result.insert(0, stack.pop());
            }
            return result.toString();
        }
    }

    public static void main(String[] args) {
        SimplifyPath path = new SimplifyPath();
        println(path.simplifyPath("/../"));
        println(path.simplifyPath("/a/./b/../../c/"));
        println(path.simplifyPath("/home//foo/"));
        println(path.simplifyPath("/.../"));
        println(path.simplifyPath("/."));
        println(path.simplifyPath("/"));
        println(path.simplifyPath("/..."));
        println(path.simplifyPath("/.../.."));//"/"
        println(path.simplifyPath("/.../."));
    }
}

File: src/main/java/edu/princeton/cs/algs4/leetcode/SimplifyPath.kt
package edu.princeton.cs.algs4.leetcode

/**
 * @author TuFei
 * @date 18-11-22.
 */
class SimplifyPath {
    fun simplifyPath(path: String): String {
        val stack = Stack<String>()
        var temp = ""
        for (c in path) {
            if (temp.isEmpty()) {
                temp += c
                continue
            }
            if (c == '/') {
                when (temp) {
                    "/." -> temp = c.toString()
                    "/" -> {
                        if (!stack.isEmpty()) {
                            stack.pop()
                        }
                        temp = c.toString()
                    }
                    "/" -> {

                    }
                    else -> {
                        stack.push(temp)
                        temp = c.toString()
                    }
                }
            } else {
                temp += c
            }
        }
        if (temp == "/" && !stack.isEmpty()) {
            stack.pop()
        }
        if (temp != "/" && temp != "/." && temp != "/") {
            stack.push(temp)
        }
        return if (stack.isEmpty()) {
            "/"
        } else {
            var result: String = ""
            while (!stack.isEmpty()) {
                result = stack.pop() + result
            }
            return result
        }
    }
}

fun main(args: Array<String>) {
    val path = SimplifyPath()
    println(path.simplifyPath("/../"))
    println(path.simplifyPath("/a/./b/../../c/"))
    println(path.simplifyPath("/home//foo/"))
    println(path.simplifyPath("/.../"))
    println(path.simplifyPath("/."))
    println(path.simplifyPath("/"))
    println(path.simplifyPath("/..."))
    println(path.simplifyPath("/.../.."))//"/"
    println(path.simplifyPath("/.../."))
}